OSDOCS-10211: eBPF flow rule filtering (#74622)
openshift/openshift-docs: modules/network-observability-ebpf-rule-flow-filter.adoc

// Module included in the following assemblies:
//
// network_observability/observing-network-traffic.adoc

:_mod-docs-content-type: CONCEPT
[id="network-observability-ebpf-flow-rule-filter_{context}"]

> = eBPF flow rule filter

Reviewer: and should this be `ebpf flow filter rule`? It it a typo or thats what was suggested?

Author: I got this from the upstream doc: https://github.com/netobserv/netobserv-ebpf-agent/blob/f2a65b3639d01c06092555312c2b85b43881aa5d/docs/flow_filtering.md#ebpf-flow-rule-based-filtering

The title here is `eBPF Flow Rule Based Filtering`

Author: Should it be different 🤔

Reviewer: oh then okay

> You can use rule-based filtering to control the volume of packets cached in the eBPF flow table. For example, a filter can specify that only packets coming from port 100 should be recorded. Then all packets which do not match the filter are not cached and only the ones that match are cached.

Reviewer 2: ```suggestion
You can use rule-based filtering to control the volume of packets cached in the eBPF flow table. For example, a filter can specify that only packets coming from port 100 should be recorded. Then, only the packets that match the filter are cached and the rest are not cached.
```
Just a suggestion for simplification.

Author: Good one.

> == How ingress and egress traffic filtering works

Reviewer 2: ```suggestion
== Ingress and egress traffic filtering
```
I tried to eliminate "How" based on the [IBM quick reference guide](https://spaces.redhat.com/pages/viewpage.action?spaceKey=CCSENABLE&title=Titles+and+headings).

> == How ingress and egress traffic filtering works

Reviewer 3: Your module titles need an anchor id and the `_{context}` variable. This needs to be addressed in a few places, but this is the only place I will call it out. https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-in-module-files

Author: Ah dangit I forgot that even 2nd level headers need anchor id.

> You can use rule-based filtering to refine where the eBPF agent deploys and what is collected. The rule-based filtering is helpful for controlling the volume of packets cached in the eBPF flow table. For example, a filter can specify that only packets coming from port 100 should be recorded. Then all packets which do not match the filter are dropped and only the ones that match pass.

Reviewer 4: well to decide where the agent will be deployed we have nodeSelector/tolartions etc which is irrelevant to flow filtering

Reviewer 4: packets that don't match filters aren't dropped they are just not cached in netobserv flow table so no longer visible from NOO pov

Author: > well to decide where the agent will be deployed we have nodeSelector/tolartions etc which is irrelevant to flow filtering

OK, maybe I misinterpreted what you wrote in the Jira [epic](https://issues.redhat.com/browse/NETOBSERV-1598) for this: 
# Epic Goal
This epic will help customers to fine grain where eBPF agents deploys and what is collected.

I will remove this detail. 

> packets that don't match filters aren't dropped they are just not cached in netobserv flow table so no longer visible from NOO pov

Okay, thanks I will update this detail.

> CIDR notation allows for the efficient representation of IP address ranges by combining the base IP address with a prefix length. For both ingress and egress traffic, the source IP address is first used to match filter rules configured with CIDR notation. If the source IP address isn't configured, then the destination IP is used to match filter rules configured with CIDR notation. 

Reviewer 4: what does it mean by `the source IP address isn't configure` this is the pkt srcIP no configuration here the only config here is the filter rule CIDR 
we 1st match packet srcIP against the CIDR if match we proceed if no match we check against pkt dstIP

Author: Yeah you're right here. I honestly don't know why I wrote `configure` here, I can see how it needs to be changed to `match`

> CIDR notation allows for the efficient representation of IP address ranges by combining the base IP address with a prefix length. For both ingress and egress traffic, the source IP address is first used to match filter rules configured with CIDR notation. If there is a match, then the filtering proceeds. If there is no match, then the destination IP is used to match filter rules configured with CIDR notation. 

Reviewer 3: ```suggestion
CIDR notation efficiently represents IP address ranges by combining the base IP address with a prefix length. For both ingress and egress traffic, the source IP address is first used to match filter rules configured with CIDR notation. If there is a match, then the filtering proceeds. If there is no match, then the destination IP is used to match filter rules configured with CIDR notation. 
```
Nit/suggestion: I know we try to avoid adverbs, but in this case, I think it expresses your intended meaning. BTW, great job describing CIDR notation!

> After matching either the source IP or the destination IP CIDR, you can pinpoint specific endpoints using the `peerIP` to differentiate the destination IP address of the packet. Based on the provisioned action, the flow data is either cached in the eBPF flow table or not. 

Author: I think I will keep the "the" before `peerIP` but I will add cached to the end of this sentence.

== Dashboard and metrics integrations

> When this option is enabled, the *Netobserv/Health* dashboard for *eBPF agent statistics* now has the *Filtered flows rate* view. Additionally, in *Observe* -> *Metrics* you can query `netobserv_agent_filtered_flows_total` to observe metrics with the reason in *FlowFilterAcceptCounter*, *FlowFilterNoMatchCounter* or *FlowFilterRecjectCounter*.

Reviewer 4: does adding an image from console here of flowfilter metrics help ?

Author: At the beginning of NetObserv project, the team and I discussed whether or not to add screencaptures to the product documentation due to the heavy console pieces. Ultimately we decided that screencaptures would be a lot of maintenance overhead, which is why we have no screencaps in the docs today. So TL;DR, it may help but the concern is maintenance.
